### DISCUSSION TITLE
Is Lattigo completely single-threaded by default?

### DISCUSSION THREAD

**Discussion by Tokubara** [tuneinsight/lattigo#526] [Q&A] [answered]
When I run a program, I notice using `htop` that only one core is being utilized. I'm wondering if Lattigo is entirely single-threaded by default (unless the user explicitly uses goroutines or `sync`)?

**Comment by Pro7ech** [comment, marked as answer]
Yes, and structures above the ring package are not thread safe. But light copies (that only reallocate non-thread safe buffers) can be created with the `ShallowCopy()` method.
